gnu/usr.bin/as/as.1
.\" Copyright (c) 1991, 1992 Free Software Foundation
.\" See section COPYING for conditions for redistribution
.\" $FreeBSD: src/gnu/usr.bin/as/as.1,v 1.9 1999/08/27 23:34:11 peter Exp $
.\" $DragonFly: src/gnu/usr.bin/as/Attic/as.1,v 1.2 2003/06/17 04:25:44 dillon Exp $
.TH as 1 "21 January 1992" "cygnus support" "GNU Development Tools"

.SH NAME
as \- the portable GNU assembler.

.SH SYNOPSIS
.na
.B as
.RB "[\|" \-a "\||\|" \-al "\||\|" -as\c
\&\|]
.RB "[\|" \-D "\|]"
.RB "[\|" \-f "\|]"
.RB "[\|" \-I
.I path\c
\&\|]
.RB "[\|" \-k "\|]"
.RB "[\|" \-K "\|]"
.RB "[\|" \-L "\|]"
.RB "[\|" \-o 
.I objfile\c
\&\|]
.RB "[\|" \-R "\|]"
.RB "[\|" \-v "\|]"
.RB "[\|" \-w "\|]"
.RB "[\|" \-\^\- "\ |\ " \c
.I files\c
\&\|.\|.\|.\|]

.I i960-only options:
.br
.RB "[\|" \-ACA "\||\|" \-ACA_A "\||\|" \-ACB\c
.RB "\||\|" \-ACC "\||\|" \-AKA "\||\|" \-AKB\c
.RB "\||\|" \-AKC "\||\|" \-AMC "\|]"
.RB "[\|" \-b "\|]"
.RB "[\|" \-norelax "\|]"

.I m680x0-only options:
.br
.RB "[\|" \-l "\|]"
.RB "[\|" \-mc68000 "\||\|" \-mc68010 "\||\|" \-mc68020 "\|]"
.ad b

.SH DESCRIPTION
GNU \c
.B as\c
\& is really a family of assemblers.  
If you use (or have used) the GNU assembler on one architecture, you
should find a fairly similar environment when you use it on another
architecture.  Each version has much in common with the others,
including object file formats, most assembler directives (often called
\c
.I pseudo-ops)\c
\& and assembler syntax.  

For information on the syntax and pseudo-ops used by GNU \c
.B as\c
\&, see `\|\c
.B as\c
\|' entry in \c
.B info \c
(or the manual \c
.I
.I
Using as: The GNU Assembler\c
\&).

\c
.B as\c
\& is primarily intended to assemble the output of the GNU C
compiler \c
.B gcc\c
\& for use by the linker \c
.B ld\c
\&.  Nevertheless,
we've tried to make \c
.B as\c
\& assemble correctly everything that the native
assembler would.
This doesn't mean \c
.B as\c
\& always uses the same syntax as another
assembler for the same architecture; for example, we know of several
incompatible versions of 680x0 assembly language syntax.

Each time you run \c
.B as\c
\& it assembles exactly one source
program.  The source program is made up of one or more files.
(The standard input is also a file.)

If \c
.B as\c
\& is given no file names it attempts to read one input file
from the \c
.B as\c
\& standard input, which is normally your terminal.  You
may have to type \c
.B ctl-D\c
\& to tell \c
.B as\c
\& there is no more program
to assemble.  Use `\|\c
.B \-\^\-\c
\|' if you need to explicitly name the standard input file
in your command line.

.B as\c
\& may write warnings and error messages to the standard error
file (usually your terminal).  This should not happen when \c
.B as\c
\& is
run automatically by a compiler.  Warnings report an assumption made so
that \c
.B as\c
\& could keep assembling a flawed program; errors report a
grave problem that stops the assembly.

.SH OPTIONS
.TP
.BR \-a \||\| \-al \||\| \-as
Turn on assembly listings; `\|\c
.B \-al\c
\&\|', listing only, `\|\c
.B \-as\c
\&\|', symbols
only, `\|\c
.B \-a\c
\&\|', everything.
.TP
.B \-D
This option is accepted only for script compatibility with calls to
other assemblers; it has no effect on \c
.B as\c
\&.
.TP
.B \-f
``fast''--skip preprocessing (assume source is compiler output).
.TP
.BI "\-I\ " path
Add 
.I path
to the search list for 
.B .include
directives.
.TP
.B \-k
Handle position independent code, generated by gcc -fpic.
.TP
.B \-K
Handle position independent code, generated by gcc -fPIC.
This is normally the same as -k.
.TP
.B \-L
Keep (in symbol table) local symbols, starting with `\|\c
.B L\c
\|'
.TP
.BI "\-o\ " objfile
Name the object-file output from \c
.B as
.TP
.B \-R
Fold data section into text section
.TP
.B \-v
Announce \c
.B as\c
\& version
.TP
.B \-W
Suppress warning messages
.TP
.IR "\-\^\-" "\ |\ " "files\|.\|.\|."
Source files to assemble, or standard input (\c
.BR "\-\^\-" ")"
.TP
.BI \-A var
.I
(When configured for Intel 960.)
Specify which variant of the 960 architecture is the target.
.TP
.B \-b
.I
(When configured for Intel 960.)
Add code to collect statistics about branches taken.
.TP
.B \-norelax
.I
(When configured for Intel 960.)
Do not alter compare-and-branch instructions for long displacements;
error if necessary.
.TP
.B \-l
.I
(When configured for Motorola 68000).  
.br
Shorten references to undefined symbols, to one word instead of two.
.TP
.BR "\-mc68000" "\||\|" "\-mc68010" "\||\|" "\-mc68020"
.I
(When configured for Motorola 68000).  
.br
Specify what processor in the 68000 family is the target (default 68020)

.PP
Options may be in any order, and may be
before, after, or between file names.  The order of file names is
significant.

`\|\c
.B \-\^\-\c
\|' (two hyphens) by itself names the standard input file
explicitly, as one of the files for \c
.B as\c
\& to assemble.

Except for `\|\c
.B \-\^\-\c
\|' any command line argument that begins with a
hyphen (`\|\c
.B \-\c
\|') is an option.  Each option changes the behavior of
\c
.B as\c
\&.  No option changes the way another option works.  An
option is a `\|\c
.B \-\c
\|' followed by one or more letters; the case of
the letter is important.   All options are optional.

The `\|\c
.B \-o\c
\|' option expects exactly one file name to follow.  The file
name may either immediately follow the option's letter (compatible
with older assemblers) or it may be the next command argument (GNU
standard).  

These two command lines are equivalent:
.br
.B
as\ \ \-o\ \ my\-object\-file.o\ \ mumble.s
.br
.B
as\ \ \-omy\-object\-file.o\ \ mumble.s

.SH "SEE ALSO"
.RB "`\|" as "\|'"
entry in 
.B
info\c
\&; 
.I
Using as: The GNU Assembler\c
\&;
.BR gcc "(" 1 "),"
.BR ld "(" 1 ")."

.SH HISTORY
A
.I as
command appeared in
Version 1 AT&T UNIX.

.SH COPYING
Copyright (c) 1991, 1992 Free Software Foundation, Inc.
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translations approved by the Free Software Foundation instead of in
the original English.


gnu/usr.bin/as/as.c
/* as.c - GAS main program.
   Copyright (C) 1987, 1990, 1991, 1992 Free Software Foundation, Inc.

   This file is part of GAS, the GNU Assembler.

   GAS is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   GAS is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with GAS; see the file COPYING.  If not, write to
   the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. */

/*
 * Main program for AS; a 32-bit assembler of GNU.
 * Understands command arguments.
 * Has a few routines that don't fit in other modules because they
 * are shared.
 *
 *
 *			bugs
 *
 * : initialisers
 *	Since no-one else says they will support them in future: I
 * don't support them now.
 *
 * $FreeBSD: src/gnu/usr.bin/as/as.c,v 1.7 1999/08/27 23:34:11 peter Exp $
 * $DragonFly: src/gnu/usr.bin/as/Attic/as.c,v 1.2 2003/06/17 04:25:44 dillon Exp $
 */

#include <stdio.h>
#include <string.h>

#ifdef _POSIX_SOURCE
#include <sys/types.h>	/* For pid_t in signal.h */
#endif
#include <signal.h>

#define COMMON

#include "as.h"
#include "subsegs.h"
#if __STDC__ == 1

/* This prototype for got_sig() is ansi.  If you want
   anything else, then your compiler is lying to you when
   it says that it is __STDC__.  If you want to change it,
   #ifdef protect it from those of us with real ansi
   compilers. */

#define SIGTY void

static void got_sig(int sig);
static char *stralloc(char *str);
static void perform_an_assembly_pass(int argc, char **argv);

#else /* __STDC__ */

#ifndef SIGTY
#define SIGTY int
#endif

static SIGTY got_sig();
static char *stralloc();	/* Make a (safe) copy of a string. */
static void perform_an_assembly_pass();

#endif /* not __STDC__ */

#ifdef DONTDEF
static char * gdb_symbol_file_name;
long gdb_begin();
#endif

int listing; /* true if a listing is wanted */

char *myname;		/* argv[0] */
extern const char version_string[];

int main(argc,argv)
int argc;
char **argv;
{
	int work_argc;	/* variable copy of argc */
	char **work_argv;	/* variable copy of argv */
	char *arg;		/* an arg to program */
	char a;		/* an arg flag (after -) */
	static const int sig[] = { SIGHUP, SIGINT, SIGPIPE, SIGTERM, 0};

	for (a=0;sig[a] != 0;a++)
	    if (signal(sig[a], SIG_IGN) != SIG_IGN)
		signal(sig[a], got_sig);

	myname=argv[0];
	memset(flagseen, '\0', sizeof(flagseen)); /* aint seen nothing yet */
#ifndef OBJ_DEFAULT_OUTPUT_FILE_NAME
#define OBJ_DEFAULT_OUTPUT_FILE_NAME "a.out"
#endif /* OBJ_DEFAULT_OUTPUT_FILE_NAME */
	out_file_name = OBJ_DEFAULT_OUTPUT_FILE_NAME;

	symbol_begin();		/* symbols.c */
	subsegs_begin();		/* subsegs.c */
	read_begin();			/* read.c */
	md_begin();			/* MACHINE.c */
	input_scrub_begin();		/* input_scrub.c */
#ifdef DONTDEF
	gdb_symbol_file_name = 0;
#endif
	/*
	 * Parse arguments, but we are only interested in flags.
	 * When we find a flag, we process it then make it's argv[] NULL.
	 * This helps any future argv[] scanners avoid what we processed.
	 * Since it is easy to do here we interpret the special arg "-"
	 * to mean "use stdin" and we set that argv[] pointing to "".
	 * After we have munged argv[], the only things left are source file
	 * name(s) and ""(s) denoting stdin. These file names are used
	 * (perhaps more than once) later.
	 */
	/* FIXME-SOMEDAY this should use getopt. */
	work_argc = argc-1;		/* don't count argv[0] */
	work_argv = argv+1;		/* skip argv[0] */
	for (;work_argc--;work_argv++) {
		arg = * work_argv;	/* work_argv points to this argument */

		if (*arg != '-')		/* Filename. We need it later. */
		    continue;	/* Keep scanning args looking for flags. */
		if (arg[1] == '-' && arg[2] == 0) {
			/* "--" as an argument means read STDIN */
			/* on this scan, we don't want to think about filenames */
			* work_argv = "";	/* Code that means 'use stdin'. */
			continue;
		}
		/* This better be a switch. */
		arg ++;		/*->letter. */

		while ((a = * arg) != '\0')  {/* scan all the 1-char flags */
			arg ++;	/* arg->after letter. */
			a &= 0x7F;	/* ascii only please */
			/* if (flagseen[a])
			   as_tsktsk("%s: Flag option - %c has already been seen.", myname, a); */
			flagseen[a] = 1;
			switch (a) {

			case 'a':
				{
					int loop =1;

					while (loop) {
						switch (*arg)
						    {
						    case 'l':
							    listing |= LISTING_LISTING;
							    arg++;
							    break;
						    case 's':
							    listing |= LISTING_SYMBOLS;
							    arg++;
							    break;
						    case 'h':
							    listing |= LISTING_HLL;
							    arg++;
							    break;

						    case 'n':
							    listing |= LISTING_NOFORM;
							    arg++;
							    break;
						    case 'd':
							    listing |= LISTING_NODEBUG;
							    arg++;
							    break;
						    default:
							    if (!listing)
								listing= LISTING_DEFAULT;
							    loop = 0;
							    break;
						    }
					}
				}

				break;


			case 'f':
				break;	/* -f means fast - no need for "app" preprocessor. */

			case 'D':
				/* DEBUG is implemented: it debugs different */
				/* things to other people's assemblers. */
				break;

#ifdef DONTDEF
			case 'G':	/* GNU AS switch: include gdbsyms. */
				if (*arg)	/* Rest of argument is file-name. */
				    gdb_symbol_file_name = stralloc (arg);
				else if (work_argc) {	/* Next argument is file-name. */
					work_argc --;
					* work_argv = NULL; /* Not a source file-name. */
					gdb_symbol_file_name = * ++ work_argv;
				} else
				    as_warn("%s: I expected a filename after -G", myname);
				arg = "";	/* Finished with this arg. */
				break;
#endif

			case 'I': { /* Include file directory */

				char *temp = NULL;
				if (*arg)
				    temp = stralloc (arg);
				else if (work_argc) {
					* work_argv = NULL;
					work_argc--;
					temp = * ++ work_argv;
				} else
				    as_warn("%s: I expected a filename after -I", myname);
				add_include_dir (temp);
				arg = "";	/* Finished with this arg. */
				break;
			}

			case 'L': /* -L means keep L* symbols */
				break;

			case 'o':
				if (*arg)	/* Rest of argument is object file-name. */
				    out_file_name = stralloc (arg);
				else if (work_argc) {	/* Want next arg for a file-name. */
					* work_argv = NULL; /* This is not a file-name. */
					work_argc--;
					out_file_name = * ++ work_argv;
				} else
				    as_warn("%s: I expected a filename after -o. \"%s\" assumed.", myname, out_file_name);
				arg = "";	/* Finished with this arg. */
				break;

			case 'R':
				/* -R means put data into text segment */
				break;

			case 'v':
#ifdef	OBJ_VMS
				{
					extern char *compiler_version_string;
					compiler_version_string = arg;
				}
#else /* not OBJ_VMS */
				fprintf(stderr,version_string);
				if (*arg && strcmp(arg,"ersion"))
				    as_warn("Unknown -v option ignored");
#endif /* not OBJ_VMS */
				while (*arg) arg++;	/* Skip the rest */
				break;

			case 'W':
				/* -W means don't warn about things */
			case 'X':
				/* -X means treat warnings as errors */
			case 'Z':
				/* -Z means attempt to generate object file even after errors. */
				break;

			default:
				--arg;
				if (md_parse_option(&arg,&work_argc,&work_argv) == 0)
				    as_warn("%s: I don't understand '%c' flag.", myname, a);
				if (arg && *arg)
				    arg++;
				break;
			}
		}
		/*
		 * We have just processed a "-..." arg, which was not a
		 * file-name. Smash it so the
		 * things that look for filenames won't ever see it.
		 *
		 * Whatever work_argv points to, it has already been used
		 * as part of a flag, so DON'T re-use it as a filename.
		 */
		*work_argv = NULL; /* NULL means 'not a file-name' */
	}
#ifdef PIC
	if (flagseen['K'] || flagseen['k'])
		picmode = 1;
#endif
#ifdef DONTDEF
	if (gdb_begin(gdb_symbol_file_name) == 0)
	    flagseen['G'] = 0;	/* Don't do any gdbsym stuff. */
#endif
	/* Here with flags set up in flagseen[]. */

	perform_an_assembly_pass(argc,argv); /* Assemble it. */
#ifdef TC_I960
	brtab_emit();
#endif
	if (seen_at_least_1_file()
	    && !((had_warnings() && flagseen['Z'])
		 || had_errors() > 0)) {
		write_object_file(); /* relax() addresses then emit object file */
	} /* we also check in write_object_file() just before emit. */

	input_scrub_end();
	md_end();			/* MACHINE.c */

#ifndef NO_LISTING
	listing_print("");
#endif

#ifndef	HO_VMS
	return((had_warnings() && flagseen['Z'])
	       || had_errors() > 0);			/* WIN */
#else	/* HO_VMS */
	return(!((had_warnings() && flagseen['Z'])
		 || had_errors() > 0));			/* WIN */
#endif	/* HO_VMS */

} /* main() */


/*			perform_an_assembly_pass()
 *
 * Here to attempt 1 pass over each input file.
 * We scan argv[*] looking for filenames or exactly "" which is
 * shorthand for stdin. Any argv that is NULL is not a file-name.
 * We set need_pass_2 TRUE if, after this, we still have unresolved
 * expressions of the form (unknown value)+-(unknown value).
 *
 * Note the un*x semantics: there is only 1 logical input file, but it
 * may be a catenation of many 'physical' input files.
 */
static void perform_an_assembly_pass(argc, argv)
int argc;
char **argv;
{
	int saw_a_file = 0;
	need_pass_2		= 0;

#ifdef MANY_SEGMENTS
	unsigned int i;

	for (i= SEG_E0; i < SEG_UNKNOWN; i++)
	    {
		    segment_info[i].fix_root = 0;
	    }
	/* Create the three fixed ones */
	subseg_new (SEG_E0, 0);
	subseg_new (SEG_E1, 0);
	subseg_new (SEG_E2, 0);
	strcpy(segment_info[SEG_E0].scnhdr.s_name,".text");
	strcpy(segment_info[SEG_E1].scnhdr.s_name,".data");
	strcpy(segment_info[SEG_E2].scnhdr.s_name,".bss");

	subseg_new (SEG_E0, 0);
#else /* not MANY_SEGMENTS */
	text_fix_root		= NULL;
	data_fix_root		= NULL;
	bss_fix_root		= NULL;

	subseg_new (SEG_TEXT, 0);
#endif /* not MANY_SEGMENTS */

	argv++; /* skip argv[0] */
	argc--; /* skip argv[0] */
	while (argc--) {
		if (*argv) { /* Is it a file-name argument? */
			saw_a_file++;
			/* argv->"" if stdin desired, else->filename */
			read_a_source_file(*argv);
		}
		argv++; /* completed that argv */
	}
	if (!saw_a_file)
	    read_a_source_file("");
} /* perform_an_assembly_pass() */

/*
 *			stralloc()
 *
 * Allocate memory for a new copy of a string. Copy the string.
 * Return the address of the new string. Die if there is any error.
 */

static char *
    stralloc (str)
char *	str;
{
	register char *	retval;
	register long len;

	len = strlen (str) + 1;
	retval = xmalloc (len);
	(void) strcpy(retval, str);
	return(retval);
}

#ifdef comment
static void lose() {
	as_fatal("%s: 2nd pass not implemented - get your code from random(3)", myname);
	return;
} /* lose() */
#endif /* comment */

static SIGTY
    got_sig(sig)
int sig;
{
	static here_before = 0;

	as_bad("Interrupted by signal %d", sig);
	if (here_before++)
	    exit(1);
	return((SIGTY) 0);
}

/*
 * Local Variables:
 * comment-column: 0
 * fill-column: 131
 * End:
 */

/* end of as.c */
